include/boost/geometry/formulas/karney_inverse.hpp

```diff
@@ -0,0 +1,904 @@
+
+public:
+    typedef result_inverse<CT> result_type;
+    static CT constexpr c0 = 0;
+    static CT constexpr c0_001 = 0.001;
+    static CT constexpr c0_01 = 0.01;
+    static CT constexpr c0_1 = 0.1;
+    static CT constexpr c0_5 = 0.5;
+    static CT constexpr c1 = 1;
+    static CT constexpr c2 = 2;
+    static CT constexpr c3 = 3;
+    static CT constexpr c4 = 4;
+    static CT constexpr c6 = 6;
+    static CT constexpr c8 = 8;
+    static CT constexpr c10 = 10;
+    static CT constexpr c16 = 16;
+    static CT constexpr c20 = 20;
+    static CT constexpr c90 = 90;
+    static CT constexpr c180 = 180;
+    static CT constexpr c200 = 200;
+    static CT constexpr c1000 = 1000;
+
+    template <typename T1, typename T2, typename Spheroid>
+    static inline result_type apply(T1 const& lo1,
+                                    T1 const& la1,
+                                    T2 const& lo2,
+                                    T2 const& la2,
+                                    Spheroid const& spheroid)
+    {
+        result_type result;
+
+        CT lat1 = la1;
+        CT lat2 = la2;
+
+        CT lon1 = lo1;
+        CT lon2 = lo2;
+
+        CT const a = CT(get_radius<0>(spheroid));
+        CT const b = CT(get_radius<2>(spheroid));
+        CT const f = formula::flattening<CT>(spheroid);
+        CT const one_minus_f = c1 - f;
+        CT const two_minus_f = c2 - f;
+
+        CT const tol0 = std::numeric_limits<CT>::epsilon();
+        CT const tol1 = c200 * tol0;
+        CT const tol2 = sqrt(tol0);
+
+        // Check on bisection interval.
+        CT const tol_bisection = tol0 * tol2;
+
+        CT const etol2 = c0_1 * tol2 /
+            sqrt(std::max(CT(0.001), std::abs(f)) * std::min(CT(1), CT(1) - f / CT(2)) / c2);
+
+        CT tiny = std::sqrt(std::numeric_limits<CT>::min());
+
+        CT const n = f / two_minus_f;
+        CT const e2 = f * two_minus_f;
+        CT const ep2 = e2 / math::sqr(one_minus_f);
+
+        // Compute the longitudinal difference.
+        CT lon12_error;
+        CT lon12 = math::difference_angle(lon1, lon2, lon12_error);
+
+        int lon12_sign = lon12 >= 0 ? 1 : -1;
+
+        // Make points close to the meridian to lie on it.
+        lon12 = lon12_sign * math::round_angle(lon12);
+        lon12_error = math::round_angle((c180 - lon12) - lon12_sign * lon12_error);
+
+        // Convert to radians.
+        CT lam12 = lon12 * math::d2r<CT>();
+        CT sin_lam12;
+        CT cos_lam12;
+
+        if (lon12 > c90)
+        {
+            math::sin_cos_degrees(lon12_error, sin_lam12, cos_lam12);
+            cos_lam12 *= -c1;
+        }
+        else
+        {
+            math::sin_cos_degrees(lon12, sin_lam12, cos_lam12);
+        }
+
+        // Make points close to the equator to lie on it.
+        lat1 = std::abs(lat1) > c90 ? math::NaN<CT>() : lat1;
+        lat2 = std::abs(lat2) > c90 ? math::NaN<CT>() : lat2;
```

Comment:
`NaN()` for fundamental types may return `quiet_nan` or `max` depending on the compiler. For user-defined numeric types it will return 0. So this is not a good way to check this.

But I'm confused. Why is it even needed here? AFAIU if `lat1` or `lat2` was NaN the result would be NaN as well. So why not simply return NaN here without performing all of the calculation below?

Other options would be to:
  - return some other value like numeric_limits::max (which will only work for fundamental types),
  - return result outside valid bounds like negative distance,
  - throw an exception.

In fact we should be consistent across the whole library with handling of latitudes > 90. In all geographic strategies and projections the coordinates should be treated the same way. Why not simply move the coordinate to 90 or -90 deg? This would also handle cases when latitude was a result of some numerical operation and e.g. is slightly greater than 90 or below -90 (like `90.0000000001`).

Reply:
The treatment of "out of bounds" latitudes is a policy decision which
should apply to the library as a whole.  GeographicLib's policy is to
allow arbitrary longitude and azimuth (with exact range reduction
preformed as necessary) and to require that latitudes lie in the range
[-90,90] deg.

The latitude range is checked by the routines which handle user input
and exceptions are thrown if needed.  With the lower level routines, out
of range latitudes are silently converted to NaNs in the expectation
that the resulting NaN will flag an obvious error to the user.

This is prefered to a strategy of defining a protocol to convert the
point to a legal one.  Note that the GeographicLib interface handles all
angles in degrees.  So a latitude like 90.0000000001 will never by
"accidentally" returned.

There's no particular reason why the policy for Boost should match was
GeographicLib does.

Reply:
@awulkiew - For the moment I have moved the coordinates to 90 deg when they are greater, and the test cases seem to pass.